labelindex/label_inheritance_index.go

```diff
@@ -110,105 +140,150 @@ func (idx *InheritIndex) DeleteSelector(id interface{}) {
+	}
+}
+
+func (idx *InheritIndex) onItemParentsUpdate(id interface{}, oldParents, newParents []*parentData) {
+	log.WithFields(log.Fields{
+		"oldParents": oldParents,
+		"newParents": newParents,
+		"id":         id,
+	}).Debug("Updating parents")
+	for _, parentData := range oldParents {
+		parentData.itemIDs.Discard(id)
+		if parentData.itemIDs.Len() == 0 {
+			parentData.itemIDs = nil
+		}
+		idx.discardParentIfEmpty(parentData.id)
+	}
+	for _, parent := range newParents {
+		if parent.itemIDs == nil {
+			parent.itemIDs = set.New()
+		}
+		parent.itemIDs.Add(id)
 	}
 }
 
 func (idx *InheritIndex) UpdateParentLabels(parentID string, labels map[string]string) {
-	idx.labelsByParentID[parentID] = labels
+	parent := idx.getOrCreateParent(parentID)
+	parent.labels = labels
 	idx.flushChildren(parentID)
 }
 
 func (idx *InheritIndex) DeleteParentLabels(parentID string) {
-	delete(idx.labelsByParentID, parentID)
+	parent := idx.getOrCreateParent(parentID)
+	parent.labels = nil
+	idx.discardParentIfEmpty(parentID)
 	idx.flushChildren(parentID)
 }
 
-func (idx *InheritIndex) UpdateParentTags(parentID string, Tags []string) {
-	idx.tagsByParentID[parentID] = Tags
+func (idx *InheritIndex) UpdateParentTags(parentID string, tags []string) {
+	parent := idx.parentDataByParentID[parentID]
+	if parent == nil {
+		parent = &parentData{}
+		idx.parentDataByParentID[parentID] = parent
+	}
+	parent.tags = tags
 	idx.flushChildren(parentID)
 }
 
 func (idx *InheritIndex) DeleteParentTags(parentID string) {
-	delete(idx.tagsByParentID, parentID)
+	parentData := idx.parentDataByParentID[parentID]
+	if parentData == nil {
+		return
+	}
+	parentData.tags = nil
+	idx.parentDataByParentID[parentID] = parentData
+	idx.discardParentIfEmpty(parentID)
 	idx.flushChildren(parentID)
 }
 
-func (idx *InheritIndex) flushChildren(parentID interface{}) {
-	idx.itemIDsByParentID.Iter(parentID, func(itemID interface{}) {
-		log.Debug("Marking child ", itemID, " dirty")
-		idx.dirtyItemIDs[itemID] = true
-	})
+func (idx *InheritIndex) flushChildren(parentID string) {
+	parentData := idx.parentDataByParentID[parentID]
+	if parentData != nil && parentData.itemIDs != nil {
+		parentData.itemIDs.Iter(func(itemID interface{}) error {
+			log.Debug("Marking child ", itemID, " dirty")
+			idx.dirtyItemIDs.Add(itemID)
+			return nil
+		})
+	}
 	idx.flushUpdates()
 }
 
 func (idx *InheritIndex) flushUpdates() {
-	for itemID := range idx.dirtyItemIDs {
+	idx.dirtyItemIDs.Iter(func(itemID interface{}) error {
 		log.Debugf("Flushing %#v", itemID)
-		itemLabels, ok := idx.labelsByItemID[itemID]
+		itemData, ok := idx.itemDataByID[itemID]
 		if !ok {
 			// Item deleted.
 			log.Debugf("Flushing delete of item %v", itemID)
 			idx.index.DeleteLabels(itemID)
 		} else {
 			// Item updated/created, re-evaluate labels.
 			log.Debugf("Flushing update of item %v", itemID)
-			combinedLabels := make(map[string]string)
-			parentIDs := idx.parentIDsByItemID[itemID]
```

Comment:
No, this is the core change of this PR.  We no longer calculate an explicit copy of the combined labels, instead we implement the `Labels` interface on `itemData`.

The actual inheritance result isn't changed by this PR.  Its still:

- explicit item labels
- profile labels
- implicit labels derived from tags

(Earliest label with a particular name wins.)